### ISSUE TITLE
tf.metrics reset_state() not called when executing in graph mode

### ISSUE THREAD

**Issue by OlavHN** [tensorflow/tensorflow#31458]
`tf.keras.metrics.*` are reset using the method `.reset_states()`, however when executed in graph mode the method isn't called and the metric isn't reset.

Example to reproduce:

```
import tensorflow as tf

print(tf.__version__)

mean = tf.metrics.Mean()

def eager_working():
    
    mean(1.)
    mean.reset_states()
    mean(2.)
    
    tf.print(mean.result())

@tf.function
def graph_not_working():
    
    mean(1.)
    mean.reset_states()
    mean(2.)
    
    tf.print(mean.result())

@tf.function
def graph_working():
    mean(1.)
    mean.total.assign(0)
    mean.count.assign(0)
    mean(2.)
    
    tf.print(mean.result())


eager_working()
mean.reset_states()
graph_not_working()
mean.reset_states()
graph_working()
```

Output:
```
2.0.0-beta1
2
1.5
2
```

**Comment by oanush** [comment]
Issue replicating with TF version 2.0beta, kindly find the [gist](https://colab.sandbox.google.com/drive/1qB1JOy9v5IW-zbgRCiMS795e2Dmj92lc#scrollTo=fuCUdASHo_t7) of colab.

**Comment by soCzech** [comment]
I can confirm the same bug in TF version 2.1. It behaves rather strangely - the first two functions work but the last does not.
```
@tf.function
def working1():
    mean.reset_states()

@tf.function
def working2():
    mean(1)

@tf.function
def not_working():
    mean(1)
    mean.reset_states()
```

**Comment by foxik** [comment]
This seems to be fixed in tf-nightly-2.2.0a20200221
